### ISSUE TITLE
Quantized Convolution Layers Operation in TF-lite

### ISSUE THREAD

**Issue by ThanasisGiak** [tensorflow/tensorflow#53370]
Hello to everyone, for academic and research purposes I am trying to understand the operation behind a quantized convolution layer in Tensorflow Lite. For this purpose, I chose EffiecientNet-lite0 model. So I downloaded pretrained EfficientNet-lite0 float32 and int8 tflite files from the official [repository](https://github.com/tensorflow/tpu/tree/master/models/official/efficientnet/lite) and run inference of these models using a sample .jpg image. Firstly, I checked model's architecture and some details using Netron tool and decided to pick first Conv2D layer as my case study.

![netron](https://user-images.githubusercontent.com/57605047/145392585-5a4393f8-5ab5-4e36-864a-9927c9782a63.png)

I started with fp32 model inference as I thought it will be simplier and used the code below for preprocess the image and for the inference of the model.
```
import tensorflow as tf
import numpy as np
import PIL.Image as Image


MEAN_RGB = 127.0
STDDEV_RGB = 128.0
CROP_PADDING = 32
IMAGE_SIZE = 224

def _decode_and_center_crop(image, image_size, resize_method=Image.BICUBIC):
    """Crops to center of image with padding then scales image_size."""
    image_width, image_height = image.size
    padded_center_crop_size = int((image_size / (image_size + CROP_PADDING)) * min(image_height, image_width))
    offset_height = ((image_height - padded_center_crop_size) + 1) // 2
    offset_width = ((image_width - padded_center_crop_size) + 1) // 2
    crop_window = [offset_width, offset_height, 
                   offset_width + padded_center_crop_size, 
                   offset_height + padded_center_crop_size]
    resized_image = image.crop(crop_window)
    resized_image = resized_image.resize((image_size, image_size), resize_method)
    return resized_image

with open('image_net_classes.txt') as f:
    lines = f.readlines()

image = Image.open('beagle.jpg')
resized_image = _decode_and_center_crop(image, IMAGE_SIZE)
resized_image = np.array(resized_image).astype(np.float32)
resized_image -= MEAN_RGB
resized_image /= STDDEV_RGB
resized_image = np.expand_dims(resized_image, axis=0)


# Load TFLite model and allocate tensors.
interpreter = tf.lite.Interpreter(model_path="efficientnet-lite0-fp32.tflite", experimental_preserve_all_tensors=True)
interpreter.allocate_tensors()

# Get input and output tensors.
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

input_shape = input_details[0]['shape']
interpreter.set_tensor(input_details[0]['index'], resized_image)

interpreter.invoke()

for t in interpreter.get_tensor_details():
  if t['index'] == 102:
    test = interpreter.get_tensor(t['index'])
print(test.shape)

output_data = interpreter.get_tensor(output_details[0]['index'])
string = str(np.argmax(output_data))

for line in lines:
    if string in line:
        print('The image is a', line)
        break
```
So after I downloaded Conv2D layer's parameters(kernel) I implented fused Relu6 Conv2D layer using simple Python and later came back to compare results and everything was working pretty good.

So the next step was to implement quantized Conv2D layer of EfficientNet-lite0-int8 and used the code below.

```
MEAN_RGB = 127.0
STDDEV_RGB = 128.0
IMAGE_SIZE = 224
scale = 0.012566016986966133
zero_point = 131

image = Image.open('beagle.jpg')
resized_image = _decode_and_center_crop(image, IMAGE_SIZE)
resized_image = np.array(resized_image).astype(np.float32)
resized_image -= MEAN_RGB
resized_image /= STDDEV_RGB
resized_image = np.expand_dims(resized_image, axis=0)
resized_image = resized_image / scale + zero_point
resized_image = np.array(resized_image).astype(np.uint8)

interpreter = tf.lite.Interpreter(model_path="efficientnet-lite0-int8.tflite", experimental_preserve_all_tensors=True)
interpreter.allocate_tensors()

# Get input and output tensors.
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

input_shape = input_details[0]['shape']
interpreter.set_tensor(input_details[0]['index'], resized_image)


interpreter.invoke()
for t in interpreter.get_tensor_details():
  if t['index'] == 102:
    test = interpreter.get_tensor(t['index'])
print(test.shape)

output_data = interpreter.get_tensor(output_details[0]['index'])


for line in lines:
    if string in line:
        print('The image is a', line)
        break
```
I also studied this [paper](https://arxiv.org/pdf/1712.05877.pdf) that provided this Conv layer's implemetation [here](https://github.com/tensorflow/tensorflow/blob/4952f981be07b8bf508f8226f83c10cdafa3f0c4/tensorflow/contrib/lite/kernels/internal/reference/reference_ops.h#L248-L314). So in my understanding, the quantized Convolution Operation is the same as the full precision one but you have also to take into account offsets and scales. 

I was able to extract input,output,kernel and biases scales and offsets and managed to transform double multipliers to quantized multipliers and right shift using fuctions defined [here](https://github.com/tensorflow/tensorflow/blob/4952f981be07b8bf508f8226f83c10cdafa3f0c4/tensorflow/contrib/lite/kernels/internal/quantization_util.cc) to be able to do the needed operation below.
```
acc = MultiplyByQuantizedMultiplierSmallerThanOne(
              acc, output_multiplier, output_shift);
```

That I assume is a per-axis operation. **So my question here, is that for this operation the multiplier we use is a quantized multiplier that equals Mo = (Sinput * Skernel) / Soutput or it's something else?**

So after I wrote a Python implementation of the above quantized Con2D layer and checked the results using
```
for t in interpreter.get_tensor_details():
  if t['index'] == 102:
    test = interpreter.get_tensor(t['index'])
```
there was a big deviation. Firstly, I thought that the above layer was a fused Relu6 Conv2D layer so I was expecting output tensor's values to be between 0 and 6 but that was not the case.

**Could you please provide me a more detailed description of a quantized fused Relu6 Conv2D layer's operation?**

### Parameters defined

The sample image I used for inference.
![beagle](https://user-images.githubusercontent.com/57605047/145397592-f2d5b484-d91e-4dbd-be53-93f1135a7844.jpg)

I am using Google Collaboratory to run the above code snippets.

**Tensorflow Version:** 2.7.0
**Python version:**  3.7.12
**Numpy version:** 1.19.5

**Comment by tilakrayal** [comment]
@ThanasisGiak ,
On running the given code snippet, I am facing an different error. Please find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/7dcc05e1d0a6b2307583254e8d2813ad/untitled138.ipynb).Also please, fill issue [template](https://github.com/tensorflow/tensorflow/issues/new/choose) and the tensorflow version you are using.

**Comment by ThanasisGiak** [comment]
Hello @tilakrayal , thanks a lot for your immediate response. I forgot to attach [image_net_classes.txt](https://github.com/tensorflow/tensorflow/files/7691512/image_net_classes.txt) file. I also filled Tensorflow, Numpy and Python version I am using.

**Comment by ThanasisGiak** [comment]
After, studying carefully  Tensorflow's github repository I found kernel_util.cc file and [CalculateActivationRangeUint8](https://github.com/tensorflow/tensorflow/blob/4952f981be07b8bf508f8226f83c10cdafa3f0c4/tensorflow/contrib/lite/kernels/kernel_util.cc#L41) function. So using this function, I managed to understand why quantized fused ReLu6 Conv2D layer's output tensor is not clipped between [0, 6] but between [-128, 127] values. For the record, I managed to implement a Conv2D layer's operation in Python with some simple steps. 

- Firstly, you got to take layer's parameters(kernel, bias, scales, offsets) using _interpreter.get_tensor_details()_  command and calculate _output_multiplier_  using [GetQuantizedConvolutionMultipler](https://github.com/tensorflow/tensorflow/blob/4952f981be07b8bf508f8226f83c10cdafa3f0c4/tensorflow/contrib/lite/kernels/kernel_util.cc#L22) and [QuantizeMultiplierSmallerThanOne](https://github.com/tensorflow/tensorflow/blob/4952f981be07b8bf508f8226f83c10cdafa3f0c4/tensorflow/contrib/lite/kernels/internal/quantization_util.cc#L25) functions. 
- After that, subtract input offset from the input layer before padding it and implement a simple convolution. 
- Later, you need to use [MultiplyByQuantizedMultiplierSmallerThanOne](https://github.com/tensorflow/tensorflow/blob/4952f981be07b8bf508f8226f83c10cdafa3f0c4/tensorflow/contrib/lite/kernels/internal/reference/reference_ops.h#L36) function that uses [SaturatingRoundingDoublingHighMul](https://github.com/google/gemmlowp/blob/master/fixedpoint/fixedpoint.h#L340) and [RoundingDivideByPOT](https://github.com/google/gemmlowp/blob/master/fixedpoint/fixedpoint.h#L368) from _gemmlowp/fixedpoint.h_ library. 
- Finally, add _output_offset_ to the result and clip it using the values taken from _CalculateActivationRangeUint8_ function.

**Comment by mrtpk123** [comment]
Hello @ThanasisGiak,
I'm also trying to learn to implement quantized convolution. I have been trying to find a resource (a blog, a book, or anything) on this matter and stumbled on this issue. Can you please share the resources that you have found helpful? Any guidance on this is deeply appreciated. 
Thank you.

**Comment by ThanasisGiak** [comment]
Hello @mrtpk123 ,
Firstly, a good paper to understand the basic operations behind quantized Tensorflow layer's operations is [Quantization and Training of Neural Networks for Efficient Integer-Arithmetic-Only Inference](https://arxiv.org/pdf/1712.05877.pdf).
After that for the implementation, you have to study in detail Tensorflow's [repo](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/kernels) and follow the steps I mentioned above. Be careful, because for the third step some layer's use [Single-rounding MultiplyByQuantizedMultiplier function](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/internal/common.h#L144) and other layers use [Double-rounding MultiplyByQuantizedMultiplier function](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/internal/common.h#L236).

**Comment by mrtpk123** [comment]
Thank you @ThanasisGiak
